### PR TITLE
Extend CLI daemon based features

### PR DESCRIPTION
**Pull request by hidmic**
Connected to #419. This pull request:

- Augments daemon support for the ROS graph API. 
- Adds a `--no-daemon` option for `NodeStrategy` client code to use and allow users to NOT use nor spawn a daemon.

### CODE REVIEW

**Comment by hidmic** [comment]
Running CI up to `sros2`:

- Linux [Build Status](https://ci.ros2.org/job/ci_linux/8818/)
- Arch [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/4633/)
- macOS [Build Status](http://ci.ros2.org/job/ci_osx/7232/)
- Windows [Build Status](http://ci.ros2.org/job/ci_windows/8735/)

File: ros2cli/ros2cli/node/strategy.py [hunk 1]
```diff
@@ -23,10 +23,12 @@
 class NodeStrategy:
 
     def __init__(self, args):
-        if is_daemon_running(args):
+        use_daemon = not getattr(args, 'no_daemon', False)
```
Comment (dirk-thomas) [inline ros2cli/ros2cli/node/strategy.py]: Why use `getattr` here instead of `args.no_daemon`?

File: ros2cli/ros2cli/node/strategy.py [hunk 2]
```diff
@@ -43,3 +45,7 @@ def __exit__(self, exc_type, exc_value, traceback):
 def add_arguments(parser):
     add_daemon_node_arguments(parser)
     add_direct_node_arguments(parser)
+    parser.add_argument(
+        '--no-daemon', action='store_true', default=False,
```
Comment (dirk-thomas) [inline ros2cli/ros2cli/node/strategy.py]: Nitpick: `default=False` is not necessary. The default `default` value is `None` which would be fine.

File: ros2cli/ros2cli/node/strategy.py (hunk 1 above)
Comment (hidmic) [reply ros2cli/ros2cli/node/strategy.py]: A CLI implementation may not (yet) add the argument. Using `getattr` helps prevent failures -- just like it is done in [`DirectNode`](https://github.com/ros2/ros2cli/blob/b5349eefc6206e733f082d5de56d91583a0f70ce/ros2cli/ros2cli/node/direct.py#L34-L46) already.

File: ros2cli/ros2cli/node/strategy.py (hunk 2 above)
Comment (hidmic) [reply ros2cli/ros2cli/node/strategy.py]: True, coercion will do the trick. See https://github.com/ros2/ros2cli/pull/420/commits/040646185ccf7c914875658f8c66a55e53982b88.

File: ros2cli/ros2cli/daemon/__init__.py [hunk 3]
```diff
@@ -69,6 +69,14 @@ def main(*, script_name='_ros2_daemon', argv=None):
...
             server.register_function(
                 _print_invoked_function_name(node.get_service_names_and_types))
+            server.register_function(
+                _print_invoked_function_name(node.get_publisher_names_and_types_by_node))
+            server.register_function(
+                _print_invoked_function_name(node.get_subscriber_names_and_types_by_node))
+            server.register_function(
+                _print_invoked_function_name(node.get_service_names_and_types_by_node))
+            server.register_function(
+                _print_invoked_function_name(node.get_client_names_and_types_by_node))
```
Comment (ivanpauno) [inline ros2cli/ros2cli/daemon/__init__.py]: `count_publishers` and `count_subscribers` (which are also graph related), should be added too.

**Comment by hidmic** [comment]
> A follow up replacing `DirectNode` by `NodeStrategy` where possible would be nice.

That's coming next.

File: ros2cli/ros2cli/daemon/__init__.py (hunk 3 above)
Comment (hidmic) [reply ros2cli/ros2cli/daemon/__init__.py]: True, and I'll need it for `ros2 topic info`. Added in https://github.com/ros2/ros2cli/pull/420/commits/5a1c3136ecf8099f889ad9fc4c9edb317c94edf1.

**Comment by hidmic** [comment]
One last sanity check:

- Linux [Build Status](https://ci.ros2.org/job/ci_linux/8826/)

**Comment by nuclearsandwich** [comment]
I think these failures:
  * [sros2.test.sros2.commands.security.verbs/test_generate_policy/test_generate_policy_topics](https://ci.ros2.org/view/nightly/job/nightly_linux_release/1386/testReport/sros2.test.sros2.commands.security.verbs/test_generate_policy/test_generate_policy_topics/)
  * [sros2.test.sros2.commands.security.verbs/test_generate_policy/test_generate_policy_services](https://ci.ros2.org/view/nightly/job/nightly_linux_release/1386/testReport/sros2.test.sros2.commands.security.verbs/test_generate_policy/test_generate_policy_services/)

are occurring because https://github.com/ros2/sros2/pull/173 was merged without this PR. Is there anything preventing this from being merged?

**Comment by hidmic** [comment]
Nope, it can be merged now.

**Pull request merged**